keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,980 @@
+suggest that maps storing allocations will perform well under the number of
+nodes we expect.
+
+#### Startup
+
+-   Fetch list of `ClusterCIDRConfig` and build internal data structure
+-   If they are set, read the `--cluster-cidr` and `--node-cidr-mask-size` flags
+    and attempt to create `ClusterCIDRConfig` with the name
+    "created-from-flags". This will be used down the line for migrating users to
+    the new allocator.
+    -   In the dual-stack case, the flags `--node-cidr-mask-size-ipv4` and
+        `--node-cidr-mask-size-ipv6` are used instead, they will also be used as
+        necessary.
+-   Fetch list of `Node`s. Check each node for `PodCIDRs`
+    -   If `PodCIDR` is set, mark the allocation in the internal data structure
+        and store this association with the node.
+    -   If `PodCIDR` is set, but is not part of one of the tracked
+        `ClusterCIDRConfig`, emit a K8s event but do nothing.
+    -   If `PodCIDR` is not set, save Node for allocation in the next step.
+        After processing all nodes, allocate ranges to any nodes without Pod
+        CIDR(s) [Same logic as Node Added event]
+
+#### Reconciliation Loop
+
+This go-routine will watch for cleanup operations and failed allocations and
+continue to try them in the background.
+
+For example if a Node can't be allocated a PodCIDR, it will be periodically
+retried until it can be allocated a range or it is deleted.
+
+#### Event Watching Loops
+
+##### Node Added
+
+If the Node already has a `PodCIDR` allocated, mark the CIDRs as used.
+
+Otherwise, go through the list of `ClusterCIDRConfig`s and find ranges matching
+the node selector from each family. Attempt to allocate Pod CIDR(s) with the
+given per-node size. If that `ClusterCIDRConfig` cannot fit a node, search for
+another `ClusterCIDRConfig`.
+
+If no `ClusterCIDRConfig` matches the node, or if all matching
+`ClusterCIDRConfig`s are full, raise a K8s event and put the Node on the
+reconciliation queue (infinite retries). Upon successfully allocating CIDR(s),
+update the node object with the podCIDRs.
+
+##### Node Updated
+
+Check that its Pod CIDR(s) match internal allocation.
+
+-   If node.spec.PodCIDRs is already filled up, honor that allocation and mark
+    those ranges as allocated.
+-   If the node.spec.PodCIDRs is filled with a CIDR not from any
+    `ClusterCIDRConfig`, raise a K8sEvent.
+-   If the ranges are already marked as allocated for some other node, raise
+    another error event (there isn’t an obvious reconciliation step the
+    controller can take unilaterally).
+
+##### Node Deleted
+
+Release said Node’s allocation from the internal data-structure.
+
+If this Node is the last one using a particular `ClusterCIDRConfig` that has
+been slated for deletion, trigger the deletion flow again (so that the finalizer
+is removed and internal data structures are cleaned up).
+
+##### ClusterCIDRConfig Added
```

Comment:
I commented previously that we can do cross validation in the controller instead of the API rest, i.e. a new ClusterCIDR is added that is invalid because it overlaps or any other reason.
We can set the Status to "invalid" or something similar, to inform the user that is wrong, and the controller, will just omit it.
Since they are inmutable, the user will have to delete and create a new one with the correct configuration

Reply:
We *can* but I am not sure we need to.  If we *ever* want to do this, though, we need some way to mark it as "accepted" so we don't race.  E.g.  This would be bad:

* Add new invalid CIDR
* New node tries to use it
* Controller marks it as invalid

Instead:

* Add new invalid CIDR, state=pending
* New node tries to use it, but can't because state = pending
* Controller marks it as invalid

Again, I am not really sure we NEED to do this - we should think thru what errors it would likely catch.

Reply:
I was assuming nodes only will use cidrs with status active

> ClusterCIDRConfigStatusType = "active"

Is not similar on how pvc/pv works?

How do we "protect" the cluster then from overlapping or duplicate cidrs?

Reply:
What would make a CIDR "invalid" ?  They are explicitly allowed to overlap (so different nodes can get different sizes from the same range).

Reply:
ah,  I wasn't considering that overlapping was allowed, you are right, that was my argument for the cross validation ... 
from the top of my head, a conflict with the service ip range (subnet equal to or included into) or wrong ip families (IPv4 only cluster and IPv6 CIDR) are the only ones that can cause problems

Reply:
I think those are both interesting problems.

1. Overlapping with the service range. The current IPAM controller takes a serviceCIDR as an input and doesn't allocate anything there. Should we continue that same scheme? https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodeipam/node_ipam_controller.go#L79
2. Wong ip family is also tricky. How does Kubernetes communicate single stack vs dual stack to various components. Is there a well-known flag we can plug in to?  or maybe another Kubernetes resource?
    - Maybe we need to give this controller a flag to tell it to give every node a IPv6 address?
    - If we have both ipv4 and ipv6 objects, then we're in dual stack mode. Otherwise, it's a single stack mode.

Reply:
Once you have multiple Service CIDRs and multiple Pod CIDRs, does it make sense to try to prevent overlaping?  Especially if both of those operating modes are optional?  I guess that could argue for a single IPAM admission check that understands the different modes and prevents collisions.  Doesn't sound TOO hard?

> If we have both ipv4 and ipv6 objects, then we're in dual stack mode. Otherwise, it's a single stack mode.

I like this.